repl/repl.go

```diff
@@ -0,0 +1,178 @@
+package repl
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/spy16/sabre"
+)
+
+// New returns a new instance of REPL with given sabre Scope. Option values
+// can be used to configure REPL input, output etc.
+func New(scope sabre.Scope, opts ...Option) *REPL {
+	repl := &REPL{
+		scope:            scope,
+		currentNamespace: func() string { return "" },
+	}
+
+	if ns, ok := scope.(NamespacedScope); ok {
+		repl.currentNamespace = ns.CurrentNS
+	}
+
+	for _, option := range withDefaults(opts) {
+		option(repl)
+	}
+
+	return repl
+}
+
+// NamespacedScope can be implemented by sabre.Scope implementations to allow
+// namespace based isolation (similar to Clojure). REPL will call CurrentNS()
+// method to get the current Namespace and display it as part of input prompt.
+type NamespacedScope interface {
+	CurrentNS() string
+}
+
+// REPL implements a read-eval-print loop for a generic Runtime.
+type REPL struct {
+	scope            sabre.Scope
+	input            Input
+	output           io.Writer
+	mapInputErr      ErrMapper
+	currentNamespace func() string
+	newReader        ReaderFactory
+
+	banner      string
+	prompt      string
+	multiPrompt string
+}
+
+// Input implementation is used by REPL to read user-input. See WithInput()
+// REPL option to configure an Input.
+type Input interface {
+	SetPrompt(string)
+	Readline() (string, error)
+}
+
+// Loop starts the read-eval-print loop. Loop runs until context is cancelled
+// or input stream returns an irrecoverable error (See WithInput()).
+func (repl *REPL) Loop(ctx context.Context) error {
+	repl.printBanner()
+	repl.setPrompt(false)
+
+	if repl.scope == nil {
+		return errors.New("scope is not set")
+	}
+
+	for {
```

Comment:
Minor detail:  you can get rid of the select statement by writing:

```go
for ctx.Err() == nil {
    // read, eval, print ...
}
```

It removes one level of nesting, which is always nice.

Reply:
Yes. I keep forgetting when there is `default` in `select` i don't really need to use `ctx.Done()`